<a href="https://colab.research.google.com/github/owencastleman/Data-219-Project/blob/main/3_Machine_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Machine Learning

# Objective:


*   make a model that best predicts the type of crash in new york city
*   utilize different tools to make the most accurate model
*   produce a confusion matrix that displays the success of the model





In [ ]:
import pandas as pd
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import make_column_transformer
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
df_clean = pd.read_csv('https://raw.githubusercontent.com/owencastleman/Data-219-Project/main/Clean%20CSV%20(1).csv')
df_clean['Month of Crash'] = df_clean['Date of Crash'].str.split('/', expand = True)[0]

In [ ]:
X_train = df_clean[['VEHICLE TYPE CODE 2', 'Street', 'Neighborhood']]
Y_train = df_clean['Cause of Crash']
pipeline = make_pipeline(
    OneHotEncoder(handle_unknown='ignore'),
    KNeighborsClassifier(n_neighbors=30)
)
scores = cross_val_score(
    pipeline,
    X=X_train,
    y=Y_train,
    scoring='f1_macro',
    cv=5
)
scores.mean()

0.5862621007233942

In [ ]:
from sklearn.metrics import confusion_matrix

pipeline.fit(X_train, Y_train)
y_train_ = pipeline.predict(X_train)
confusion_matrix(Y_train, y_train_)
pd.DataFrame(
    confusion_matrix(Y_train, y_train_),
    index=pipeline.classes_,
    columns=pipeline.classes_
)

Backing Unsafely  \
Backing Unsafely                           4066   
Failure to Yield Right-of-Way              1565   
Following Too Closely                      1840   

                               Failure to Yield Right-of-Way  \
Backing Unsafely                                        2147   
Failure to Yield Right-of-Way                          10702   
Following Too Closely                                   3764   

                               Following Too Closely  
Backing Unsafely                                1377  
Failure to Yield Right-of-Way                   3204  
Following Too Closely                           7672

After using many different combinations of features, the best model that we came up with using KNeighborsClassifier was 58% accurate under f1 macro scoring.

In [ ]:
from sklearn.tree import DecisionTreeClassifier
pipeline = make_pipeline(
    OneHotEncoder(handle_unknown='ignore'),
    DecisionTreeClassifier(random_state=0)
)
scores = cross_val_score(
    pipeline,
    X=X_train,
    y=Y_train,
    scoring='f1_macro',
    cv=5
)
scores.mean()

0.5792019116634262

After doing the DecisionTreeClassifier with the same features, KNeighborsClassifier is slightly less accurate at 57.9%.

In [ ]:
from sklearn.metrics import confusion_matrix

pipeline.fit(X_train, Y_train)
y_train_ = pipeline.predict(X_train)
confusion_matrix(Y_train, y_train_)
pd.DataFrame(
    confusion_matrix(Y_train, y_train_),
    index=pipeline.classes_,
    columns=pipeline.classes_
)

Backing Unsafely  \
Backing Unsafely                           6700   
Failure to Yield Right-of-Way              1943   
Following Too Closely                      1988   

                               Failure to Yield Right-of-Way  \
Backing Unsafely                                         568   
Failure to Yield Right-of-Way                          12339   
Following Too Closely                                   1899   

                               Following Too Closely  
Backing Unsafely                                 322  
Failure to Yield Right-of-Way                   1189  
Following Too Closely                           9389

After seeing both of these, we chose to go with the DecisionTreeClassifier. All the f1 macro is lower, the confusion matrix displays that the decision tree more correctly predicts crash types. The KNeighborsClassifier has more incorrect values spread out, but it has less correct predictions.